File: clients/apple/MultiLiveTV/Services/MacCMSClient.swift
import Foundation

enum MacCMSClient {
    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = NetworkConfig.requestTimeout
        config.timeoutIntervalForResource = NetworkConfig.requestTimeout
        return URLSession(configuration: config)
    }()

    static func fetchTypes(source: Source) async throws -> [VodTypeRaw] {
        let data = try await fetchBody(baseURL: source.url, params: ["ac": "list", "pg": "1"])
        let response = try MacCMSJSONParser.parseList(data)
        return response.types
    }

    static func catalogParams(page: Int, typeId: Int?, hours: Int? = nil) -> [String: String] {
        var params = ["ac": "detail", "pg": String(page)]
        if let typeId { params["t"] = String(typeId) }
        if let hours, hours > 0 { params["h"] = String(hours) }
        return params
    }

    static func searchParams(keyword: String, page: Int) -> [String: String] {
        [
            "ac": "detail",
            "wd": keyword,
            "pg": String(page),
        ]
    }

    static func fetchList(
        source: Source,
        page: Int,
        typeId: Int?,
        hours: Int? = nil
    ) async throws -> MacCMSListResponse {
        let data = try await fetchBody(
            baseURL: source.url,
            params: catalogParams(page: page, typeId: typeId, hours: hours)
        )
        return try MacCMSJSONParser.parseList(data)
    }

    static func fetchDetail(source: Source, ids: String) async throws -> MacCMSDetailResponse {
        let data = try await fetchBody(baseURL: source.url, params: ["ac": "detail", "ids": ids])
        return try MacCMSJSONParser.parseDetail(data)
    }

    static func search(source: Source, keyword: String, page: Int) async throws -> MacCMSListResponse {
        let data = try await fetchBody(baseURL: source.url, params: searchParams(keyword: keyword, page: page))
        return try MacCMSJSONParser.parseList(data)
    }

    private static func fetchBody(baseURL: String, params: [String: String]) async throws -> Data {
        var base = baseURL
        if !base.hasSuffix("/") { base += "/" }

        guard var components = URLComponents(string: base) else {
            throw URLError(.badURL)
        }
        var queryItems = components.queryItems ?? []
        for (key, value) in params {
            queryItems.append(URLQueryItem(name: key, value: value))
        }
        components.queryItems = queryItems

        guard let url = components.url else { throw URLError(.badURL) }

        var request = URLRequest(url: url)
        request.setValue(NetworkConfig.userAgent, forHTTPHeaderField: "User-Agent")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
        return data
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/maccms/MacCMSClient.kt
package com.heibaimiao.multilivetv.maccms

import com.heibaimiao.multilivetv.model.Source
import com.heibaimiao.multilivetv.net.HttpClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject

object MacCMSClient {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    fun catalogParams(page: Int, typeId: Int?, hours: Int? = null): Map<String, String> {
        val params = mutableMapOf("ac" to "detail", "pg" to page.toString())
        if (typeId != null) params["t"] = typeId.toString()
        if (hours != null && hours > 0) params["h"] = hours.toString()
        return params
    }

    fun searchParams(keyword: String, page: Int): Map<String, String> = mapOf(
        "ac" to "detail",
        "wd" to keyword,
        "pg" to page.toString(),
    )

    suspend fun fetchTypes(source: Source): List<VodTypeRaw> = withContext(Dispatchers.IO) {
        parseList(source.url, mapOf("ac" to "list", "pg" to "1")).types
    }

    suspend fun fetchList(
        source: Source,
        page: Int,
        typeId: Int?,
        hours: Int? = null,
    ): MacCMSListResponse = withContext(Dispatchers.IO) {
        parseList(source.url, catalogParams(page, typeId, hours))
    }

    suspend fun fetchDetail(source: Source, ids: String): MacCMSDetailResponse = withContext(Dispatchers.IO) {
        val text = HttpClient.getOrThrow(source.url, extraQuery = mapOf("ac" to "detail", "ids" to ids)).text()
        MacCMSJsonParser.parseDetail(json.parseToJsonElement(text).jsonObject)
    }

    suspend fun search(source: Source, keyword: String, page: Int): MacCMSListResponse =
        withContext(Dispatchers.IO) {
            parseList(source.url, searchParams(keyword, page))
        }

    private fun parseList(baseUrl: String, params: Map<String, String>): MacCMSListResponse {
        val text = HttpClient.getOrThrow(baseUrl, extraQuery = params).text()
        return MacCMSJsonParser.parseList(json.parseToJsonElement(text).jsonObject)
    }
}
